File: Leetcode 2001-2100/2054/2054.py
# Leetcode 2054: Two Best Non-Overlapping Events
# https://leetcode.com/problems/two-best-non-overlapping-events/
# Solved on 2nd of November, 2025
import heapq


class Solution:
    def maxTwoEvents(self, events: list[list[int]]) -> int:
        """
        Calculates the maximum total value of two non-overlapping events.

        Args:
            events: A list of events, where each event is [start_time, end_time, value].
        Returns:
            The maximum total value of two non-overlapping events.
        """
        events.sort()

        minHeap = []
        maxResult = 0
        maxPrefixValue = 0

        for start, end, value in events:
            while minHeap and minHeap[0][0] < start:
                endedTime, endedValue = heapq.heappop(minHeap)
                maxPrefixValue = max(maxPrefixValue, endedValue)

            maxResult = max(maxResult, value)
            maxResult = max(maxResult, value + maxPrefixValue)

            heapq.heappush(minHeap, (end, value))

        return maxResult

File: Leetcode 2001-2100/2054/2054.java
// Leetcode 2054: Two Best Non-Overlapping Events
// https://leetcode.com/problems/two-best-non-overlapping-events/
// Solved on 2nd of November, 2025
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
    /**
     * Calculates the maximum sum of values from at most two non-overlapping events.
     *
     * @param events A 2D array where each inner array represents an event [start_time, end_time, value].
     * @return The maximum total value achievable from selecting at most two non-overlapping events.
     */
    public int maxTwoEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> Integer.compare(a[0], b[0]));

        int maxTotalSum = 0;
        int maxPrecedingValue = 0;
        
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));

        for (int[] event : events) {
            int start = event[0];
            int end = event[1];
            int value = event[2];

            while (!minHeap.isEmpty() && minHeap.peek()[0] < start) {
                int[] finishedEvent = minHeap.poll();
                int finishedValue = finishedEvent[1];
                maxPrecedingValue = Math.max(maxPrecedingValue, finishedValue);
            }

            maxTotalSum = Math.max(maxTotalSum, maxPrecedingValue + value);

            minHeap.offer(new int[]{end, value});
        }

        return maxTotalSum;
    }
}